File: in_Java/0599 Minimum Index Sum of Two Lists.java
class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String, Integer> hm1 = new HashMap<>();
        Map<String, Integer> hm2 = new HashMap<>();
        for(int i = 0; i < list1.length; i++){
            hm1.put(list1[i],i);
        }
        for(int i = 0; i < list2.length; i++){
            hm2.put(list2[i],i);
        }
        int cur = Integer.MAX_VALUE;
        int cursum;
        List<String> arr = new ArrayList<>();
        for(String k : hm1.keySet()){
            if(hm2.containsKey(k)){
                cursum = hm1.get(k) + hm2.get(k);
                if(cursum < cur){
                    cur = cursum;
                    arr = new ArrayList<>();
                    arr.add(k);
                }else if(cursum == cur){
                    arr.add(k);
                }else{
                    continue;
                }
            }
        }
        String[] res = new String[arr.size()];
        for(int i = 0; i < arr.size(); i++) res[i] = arr.get(i);
        return res;
    }
}

File: in_Python/0599 Minimum Index Sum of Two Lists.py
class Solution(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        d = {}
        dd = {}
        for i in range(len(list1)):
            r = list1[i]
            d[r] = i
        for i in range(len(list2)):
            r = list2[i]
            dd[r] = i
        minid = len(list1) + len(list2)
        res = []
        for k in d:         
            if k in dd:
                cursum = dd[k] + d[k]
                if cursum < minid:
                    res = []
                    minid = cursum
                    res.append(k)
                elif cursum == minid:
                    res.append(k)                   
        return res